e2e/LongHaul/service/IotHub.cs

```diff
@@ -123,8 +128,62 @@ await Task
+                        && module.ConnectionState is ClientConnectionState.Disconnected)
+                    {
+                        CancellationTokenSource source = s_onlineModuleOperations[moduleId].Item2;
+                        // Signal cancellation to all tasks on the particular device.
+                        source.Cancel();
+                        // Dispose the cancellation token source.
+                        source.Dispose();
+                        // Remove the correlated device operations and cancellation token source of the particular device from the dictionary.
+                        s_onlineDeviceOperations.TryRemove(moduleId, out _);
+                    }
+                    else if (!s_onlineModuleOperations.ContainsKey(moduleId)
+                        && module.ConnectionState is ClientConnectionState.Connected)
+                    {
+                        // For each online module, initiate a new cancellation token source.
+                        // Once the module goes offline, cancel all operations on this module.
+                        var source = new CancellationTokenSource();
+                        CancellationToken token = source.Token;
+
+                        async Task Operations()
+                        {
+                            var moduleOperations = new ModuleOperations(s_serviceClient, module.DeviceId, module.ModuleId, _logger.Clone());
+                            _logger.Trace($"Creating {nameof(ModuleOperations)} on the device: [{module.DeviceId}], module: [{module.ModuleId}]", TraceSeverity.Verbose);
+
+                            try
+                            {
+                                await Task
+                                .WhenAll(
+                                    moduleOperations.InvokeDirectMethodAsync(_logger.Clone(), token),
+                                    moduleOperations.SetDesiredPropertiesAsync("guidValue", Guid.NewGuid().ToString(), _logger.Clone(), token))
+                                .ConfigureAwait(false);
+                            }
+                            catch (OperationCanceledException)
+                            {
+                                _logger.Trace($"Operations on device: [{module.DeviceId}], module: [{module.ModuleId}] have been canceled as the device goes offline.", TraceSeverity.Information);
+                            }
+                            catch (Exception ex)
+                            {
+                                _logger.Trace($"Service app failed with exception {ex}", TraceSeverity.Error);
+                            }
+                        }
+
+                        // Passing in "Operations()" as Task so we don't need to manually call "Invoke()" on it.
+                        var operationsTuple = new Tuple<Task, CancellationTokenSource>(Operations(), source);
+                        s_onlineDeviceOperations.TryAdd(moduleId, operationsTuple);
```

Comment:
This should be `s_onlineModuleOperations`, right?
```suggestion
                        s_onlineModuleOperations.TryAdd(moduleId, operationsTuple);
```